### ISSUE TITLE
Guess your code does not correspond video

### ISSUE THREAD

**Issue by Overff** [visakhanc/eCompass#4]
Hello!
First of all I would like to say thank you for this project.
I tested your code and realized it does not work like you show in [video](https://www.youtube.com/watch?v=spmMMIMboPY).

If I change Roll from 0 to -15 azimuth(heading) will change from 272 to 336 (+64 dgr). See the attached pics. I am wondering, how it works in your video?.. Guess your code* does not correspond the video.
*your code:
```
/* Calculate pitch and roll, in the range (-pi,pi) */
		pitch = atan2((double)-acc_x, sqrt((long)acc_z*(long)acc_z + (long)acc_y*(long)acc_y));
		roll = atan2((double)acc_y, sqrt((long)acc_z*(long)acc_z  + (long)acc_x*(long)acc_x));

		/* Calculate Azimuth:
		 * Magnetic horizontal components, after compensating for Roll(r) and Pitch(p) are:
		 * X_h = X*cos(p) + Y*sin(r)*sin(p) + Z*cos(r)*sin(p)
		 * Y_h = Y*cos(r) - Z*sin(r)
		 * Azimuth = arcTan(Y_h/X_h)
		 */
		X_h = (double)mag_x*cos(pitch) + (double)mag_y*sin(roll)*sin(pitch) + (double)mag_z*cos(roll)*sin(pitch);
		Y_h = (double)mag_y*cos(roll) - (double)mag_z*sin(roll);
		azimuth = atan2(Y_h, X_h);
		if(azimuth < 0) {	/* Convert Azimuth in the range (0, 2pi) */
			azimuth = 2*M_PI + azimuth;
		}
```

Although my setup a little bit different, but I believe it has to work in the same way. But it does not.
My setup - MPU9250 with integrated magnetometer AK8963.
Gyro, Acc and Mag are calibrated. Azimuth (heading) calculates fine if the MPU keeps in horizontal position.

I simplified  your code and adopted it for my setup. In general it is the same!
  ```
float mag_x, mag_y, mag_z;
  float acc_x, acc_y, acc_z;
  float roll;
  float pitch;
  float azimuth;
  float X_h, Y_h;
  

  mag_x = Mag.calibrated_mx; //calibrated data
  mag_y = Mag.calibrated_my; //calibrated data
  mag_z = Mag.calibrated_mz; //calibrated data
  
  acc_x = AccX; //calibrated data
  acc_y = AccY; //calibrated data
  acc_z = AccZ; //calibrated data


    pitch = atan2(-acc_x, sqrt(acc_z*acc_z + acc_y*acc_y));
    roll = atan2(acc_y, sqrt(acc_z*acc_z  + acc_x*acc_x));
   
    X_h = mag_x*cos(pitch) + mag_y*sin(roll)*sin(pitch) + mag_z*cos(roll)*sin(pitch);
    Y_h = mag_y*cos(roll) - mag_z*sin(roll);
    
    azimuth = atan2(Y_h, X_h); 
    if(azimuth < 0) { /* Convert Azimuth in the range (0, 2pi) */
      azimuth = 2 * PI + azimuth;
    }

Serial.println(azimuth*RAD_TO_DEG); //get in degree
```
Screenshot from your video
![image2](https://github.com/visakhanc/eCompass/assets/105443564/7b69a62f-7c06-412a-93ac-cf9894cc88c1)
(after changing Roll) Heading = 336 R=-15, P=0
![image0](https://github.com/visakhanc/eCompass/assets/105443564/33771326-5788-44ee-abd3-72aaf76dbd0d)
(before changing Roll) Heading = 272 R=0, P=0
![image1](https://github.com/visakhanc/eCompass/assets/105443564/7a9eec4e-4ff8-42c1-9a15-99144a1bdb14)
Heading = 359 R=0, P=0
![image2](https://github.com/visakhanc/eCompass/assets/105443564/2ff40083-bc31-4279-a2c5-5b7af5022481)

**UPDATE**
Finally I coped with the problem. Now it works. But as well as I use 3 filters in order to smooth final azimuth.
On top of that I realized what so important to check an alignment between axes of mag, gyro and acc. In my case where two misalignments:
1. Magnetometer X doesn't correspond to Accelerometer X and Gyroscope X so I reordered register readings:
    Mag.raw_Y = (HXH << 8) | HXL;  // Y axis
    Mag.raw_X = (HYH << 8) | HYL;  // X axis
    Mag.raw_z = (HZH << 8) | HZL; // Z axis 
I have no idea that the heck is that. Probably my cheep is fake. I have already bought few fake cheeps.
2. When I raised the sensor nose (increase pitch) the pitch had negative values. But seems to be positive values. So I took it into account too.
the best way to check it - https://danceswithcode.net/engineeringnotes/rotations_in_3d/demo3D/rotations_in_3d_tool.html

So, In my case works these formulas:
```
     X_h = mag_x*cos(pitch_rad) - mag_y*sin(roll_rad)*sin(pitch_rad) + mag_z*cos(roll_rad)*sin(pitch_rad);
     Y_h = mag_y*cos(roll_rad) + mag_z*sin(roll_rad);
  
    azimuth = atan2(Y_h, X_h); //in radians
   
    if(azimuth < 0) { /* Convert Azimuth in the range (0, 2pi) */
    azimuth = 2 * PI + azimuth;
    }
```
I hope this information is useful

**Comment by brightproject** [comment]
> I tested your code and realized it does not work like you show in [video](https://www.youtube.com/watch?v=spmMMIMboPY).

Yes, the code is simply a typical implementation of tilt compensation.
In each specific case, you need to carefully and thoroughly understand the axes and their directions.

> https://www.pololu.com/file/0J434/LSM303DLH-compass-app-note.pdf

With the `ICM456868` sensor, I also had to change the code to this:
```c++
// Calculate pitch and roll, in the range (-pi,pi) 
pitch = atan2((double)-vec_raw.a_Y, sqrt((long)vec_raw.a_Z*(long)vec_raw.a_Z + (long)vec_raw.a_X*(long)vec_raw.a_X)); 
roll = atan2((double)vec_raw.a_X, sqrt((long)vec_raw.a_Z*(long)vec_raw.a_Z + (long)vec_raw.a_Y*(long)vec_raw.a_Y)); 

// Correction of magnetometer values ​​taking into account tilt 
double X_h = -Mxyz[1] * cos(pitch) - Mxyz[2] * sin(pitch); 
double Y_h = -Mxyz[1] * sin(roll) * sin(pitch) + (-Mxyz[0]) * cos(roll) + Mxyz[2] * sin(roll) * cos(pitch);

// Calculate the heading angle based on the corrected magnetometer values
azimuth_tilt = -atan2(Y_h, X_h);

// Convert the heading angle to the range from 0 to 2π
if (azimuth_tilt < 0) {
azimuth_tilt += 2 * M_PI;
}

// azimuth_tilt = magSensor.getAzimuth360calib(X_h, Y_h);

pitch *= 180.0 / PI;
roll *= 180.0 / PI;
azimuth_tilt *= 180.0 / PI;
``` 

Thanks for the [link](https://danceswithcode.net/engineeringnotes/rotations_in_3d/demo3D/rotations_in_3d_tool.html) I haven't seen that resource before.

> But as well as I use 3 filters in order to smooth final azimuth.

And if you could share the azimuth filter, like you wrote here, that would be great🙂
